File: exams/frmBestStudentSubject.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace exams
{
    public partial class frmBestStudentSubject
    {
        public frmBestStudentSubject()
        {
            InitializeComponent();
            _btnPrint.Name = "btnPrint";
            _btnCancel.Name = "btnCancel";
        }

        private void frmBestStudentSubject_Load(object sender, EventArgs e)
        {
            resultDGV.Rows.Clear();
            resultDGV.Columns.Clear();
            if (!publicSubsNFunctions.connect())
            {
                Close();
            }
            else
            {
                using (new DevExpress.Utils.WaitDialogForm("Please Wait ..."))
                {
                    getSchoolSubjects();
                    if (!ExamFunctions.getGradingScheme())
                    {
                        Interaction.MsgBox("The " + publicSubsNFunctions.gradingType + " Grading Scheme For Class " + publicSubsNFunctions.selectedClass + " Term " + publicSubsNFunctions.selectedTerm + " Year " + publicSubsNFunctions.selectedYear + " Does Not Exist In The Database. Please Add It To Proceed Or Select Another Grading Scheme");
                        return;
                    }

                    ExamFunctions.initialzeGradePoints();
                    var argdgv = resultDGV;
                    reorderColumns(ref argdgv, false);
                    resultDGV = argdgv;
                    publicSubsNFunctions.students.Clear();
                    string argq = "select admin_no, student_name, gender, class, stream from students where class = '" + publicSubsNFunctions.selectedClass + "';";
                    if (publicSubsNFunctions.qread(ref argq))
                    {
                        if (publicSubsNFunctions.dbreader.RecordsAffected > 0)
                        {
                            while (publicSubsNFunctions.dbreader.Read())
                                publicSubsNFunctions.students.Add(new Tuple<long, string, string, string, string>(Conversions.ToLong(publicSubsNFunctions.dbreader["admin_no"]), Conversions.ToString(publicSubsNFunctions.dbreader["student_name"]), Conversions.ToString(publicSubsNFunctions.dbreader["gender"]), Conversions.ToString(publicSubsNFunctions.dbreader["class"]), Conversions.ToString(publicSubsNFunctions.dbreader["stream"])));
                        }
                    }

                    string query = string.Empty;
                    short examTotal;
                    short totalMarks;
                    var rowValues = new List<string>();
                    string studGrade = string.Empty;
                    string studPoints = string.Empty;
                    foreach (var currentSubject in publicSubsNFunctions.schoolSubjects)
                    {
                        publicSubsNFunctions.subject = currentSubject;
                        var argdgv1 = computeDVG;
                        reorderColumns(ref argdgv1);
                        computeDVG = argdgv1;
                        for (int i = 0, loopTo = publicSubsNFunctions.students.Count - 1; i <= loopTo; i++)
                        {
                            totalMarks = 0;
                            for (int j = 0, loopTo1 = publicSubsNFunctions.selectedExams.Count - 1; j <= loopTo1; j++)
                            {
                                examTotal = Conversions.ToShort(publicSubsNFunctions.get_total_mark(publicSubsNFunctions.selectedExams[j].Item1, publicSubsNFunctions.selectedExams[j].Item3, publicSubsNFunctions.selectedExams[j].Item4));
                                query = "select COALESCE(((" + publicSubsNFunctions.subject + "/" + examTotal.ToString() + ") * " + publicSubsNFunctions.selectedExams[j].Item2 + "), 0) as total from exam_results where examination = '" + publicSubsNFunctions.selectedExams[j].Item1 + "' and admno = " + publicSubsNFunctions.students[i].Item1.ToString() + " and term = '" + publicSubsNFunctions.selectedExams[j].Item3 + "' and year = '" + publicSubsNFunctions.selectedExams[j].Item4.ToString() + "'";
                                if (publicSubsNFunctions.qread(ref query))
                                {
                                    if (publicSubsNFunctions.dbreader.RecordsAffected > 0)
                                    {
                                        while (publicSubsNFunctions.dbreader.Read())
                                            totalMarks = Conversions.ToShort(totalMarks + publicSubsNFunctions.dbreader["total"]);
                                    }
                                }
                            }

                            rowValues.Add("0");
                            rowValues.Add(publicSubsNFunctions.students[i].Item1.ToString());
                            rowValues.Add(publicSubsNFunctions.students[i].Item2);
                            rowValues.Add(publicSubsNFunctions.students[i].Item3);
                            rowValues.Add(totalMarks.ToString());
                            if (publicSubsNFunctions.gradingType == "SubjectBased")
                            {
                                studGrade = ExamFunctions.convertMarksToGrade(totalMarks, publicSubsNFunctions.gradingType, publicSubsNFunctions.subject);
                            }
                            else
                            {
                                studGrade = ExamFunctions.convertMarksToGrade(totalMarks, publicSubsNFunctions.gradingType);
                            }

                            rowValues.Add(studGrade);
                            studPoints = ExamFunctions.convertGradeToPoints(studGrade).ToString();
                            rowValues.Add(studPoints);
                            computeDVG.Rows.Add(rowValues.ToArray());
                            rowValues.Clear();
                        }

                        DataGridViewRow row;
                        computeDVG.Sort(computeDVG.Columns[4], System.ComponentModel.ListSortDirection.Descending);
                        if (publicSubsNFunctions.filterType == "Top")
                        {
                            if (publicSubsNFunctions.rankno + 1 < computeDVG.Rows.Count)
                            {
                                for (int i = computeDVG.Rows.Count - 1, loopTo2 = publicSubsNFunctions.rankno; i >= loopTo2; i -= 1)
                                {
                                    row = computeDVG.Rows[i];
                                    computeDVG.Rows.Remove(row);
                                }
                            }
                        }

                        if (publicSubsNFunctions.filterType == "Bottom")
                        {
                            if (publicSubsNFunctions.rankno + 1 < computeDVG.Rows.Count)
                            {
                                for (int i = computeDVG.Rows.Count - 1 - publicSubsNFunctions.rankno; i >= 0; i -= 1)
                                {
                                    row = computeDVG.Rows[i];
                                    computeDVG.Rows.Remove(row);
                                }
                            }
                        }

                        var argdgv2 = computeDVG;
                        fillResultTable(ref argdgv2, publicSubsNFunctions.subject);
                        computeDVG = argdgv2;
                    }
                }
            }
        }

        private void fillResultTable(ref DataGridView dgv, string subject)
        {
            short counter = 1;
            resultDGV.Rows.Add(subject);
            for (int k = 0, loopTo = dgv.Rows.Count - 1; k <= loopTo; k++)
            {
                resultDGV.Rows.Add(counter, dgv.Rows[k].Cells[1].Value, dgv.Rows[k].Cells[2].Value, dgv.Rows[k].Cells[3].Value, dgv.Rows[k].Cells[4].Value, dgv.Rows[k].Cells[5].Value, dgv.Rows[k].Cells[6].Value);
                counter = (short)(counter + 1);
            }
        }

        public object getSchoolSubjects()
        {
            publicSubsNFunctions.schoolSubjects.Clear();
            bool available = false;
            string argq = "select abbreviation from subjects;";
            if (publicSubsNFunctions.qread(ref argq))
            {
                if (publicSubsNFunctions.dbreader.RecordsAffected > 0)
                {
                    available = true;
                    while (publicSubsNFunctions.dbreader.Read())
                        publicSubsNFunctions.schoolSubjects.Add(Conversions.ToString(publicSubsNFunctions.dbreader["abbreviation"]));
                }
            }

            return available;
        }

        public void reorderColumns(ref DataGridView dgv, bool delete = true)
        {
            if (delete)
            {
                dgv.Rows.Clear();
                dgv.Columns.Clear();
            }

            dgv.Columns.Add("No", "No");
            dgv.Columns.Add("admno", "Adm No.");
            dgv.Columns.Add("student_name", "Name Of Student");
            dgv.Columns.Add("gender", "Gender");
            dgv.Columns.Add("marks", "marks");
            dgv.Columns.Add("grade", "grade");
            dgv.Columns.Add("point", "Points");
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void btnPrint_Click(object sender, EventArgs e)
        {
            if (resultDGV.Rows.Count > 0)
            {
                var argmyView = resultDGV;
                reporting.generateFromDataTable("Subject Wise Ranking", "From Grid", string.Empty, null, ref argmyView);
                resultDGV = argmyView;
            }
            else
            {
                Interaction.MsgBox("There are no row to print");
            }
        }
    }
}

File: exams/frmBestStudentSubject.vb
Public Class frmBestStudentSubject


    Private Sub frmBestStudentSubject_Load(sender As Object, e As EventArgs) Handles MyBase.Load


        resultDGV.Rows.Clear()
        resultDGV.Columns.Clear()

        If Not connect() Then
            Me.Close()
        Else
            Using New DevExpress.Utils.WaitDialogForm("Please Wait ...")

                getSchoolSubjects()
                If Not getGradingScheme() Then
                    MsgBox("The " + gradingType + " Grading Scheme For Class " + selectedClass + " Term " + selectedTerm + " Year " + selectedYear + " Does Not Exist In The Database. Please Add It To Proceed Or Select Another Grading Scheme")
                    Exit Sub
                End If

                initialzeGradePoints()
                reorderColumns(resultDGV, False)

                students.Clear()

                If qread("select admin_no, student_name, gender, class, stream from students where class = '" + selectedClass + "';") Then
                    If dbreader.RecordsAffected > 0 Then
                        While dbreader.Read
                            students.Add(New Tuple(Of Int64, String, String, String, String)(dbreader("admin_no"), dbreader("student_name"), dbreader("gender"), dbreader("class"), dbreader("stream")))
                        End While
                    End If
                End If

                Dim query As String = String.Empty
                Dim examTotal As Int16
                Dim totalMarks As Int16
                Dim rowValues As New List(Of String)
                Dim studGrade As String = String.Empty
                Dim studPoints As String = String.Empty


                For Each subject In schoolSubjects

                    reorderColumns(computeDVG)

                    For i = 0 To students.Count - 1

                        totalMarks = 0

                        For j = 0 To selectedExams.Count - 1
                            examTotal = get_total_mark(selectedExams(j).Item1, selectedExams(j).Item3, selectedExams(j).Item4)
                            query = "select COALESCE(((" + subject + "/" + examTotal.ToString() + ") * " + selectedExams(j).Item2 + "), 0) as total from exam_results where examination = '" + selectedExams(j).Item1 + "' and admno = " + students(i).Item1.ToString() + " and term = '" + selectedExams(j).Item3 + "' and year = '" + selectedExams(j).Item4.ToString() + "'"

                            If qread(query) Then
                                If dbreader.RecordsAffected > 0 Then
                                    While dbreader.Read
                                        totalMarks += dbreader("total")
                                    End While
                                End If
                            End If
                        Next

                        rowValues.Add("0")
                        rowValues.Add(students(i).Item1)
                        rowValues.Add(students(i).Item2)
                        rowValues.Add(students(i).Item3)
                        rowValues.Add(totalMarks)
                        If gradingType = "SubjectBased" Then
                            studGrade = convertMarksToGrade(totalMarks, gradingType, subject)
                        Else
                            studGrade = convertMarksToGrade(totalMarks, gradingType)
                        End If

                        rowValues.Add(studGrade)
                        studPoints = convertGradeToPoints(studGrade)
                        rowValues.Add(studPoints)


                        computeDVG.Rows.Add(rowValues.ToArray())
                        rowValues.Clear()
                    Next

                    Dim row As DataGridViewRow
                    computeDVG.Sort(computeDVG.Columns(4), System.ComponentModel.ListSortDirection.Descending)


                    If filterType = "Top" Then
                        If rankno + 1 < computeDVG.Rows.Count Then
                            For i = computeDVG.Rows.Count - 1 To rankno Step -1
                                row = computeDVG.Rows(i)
                                computeDVG.Rows.Remove(row)
                            Next
                        End If
                    End If

                    If filterType = "Bottom" Then

                        If rankno + 1 < computeDVG.Rows.Count Then
                            For i = (computeDVG.Rows.Count - 1) - rankno To 0 Step -1
                                row = computeDVG.Rows(i)
                                computeDVG.Rows.Remove(row)
                            Next
                        End If

                    End If


                    fillResultTable(computeDVG, subject)

                Next

            End Using
        End If

    End Sub


    Private Sub fillResultTable(ByRef dgv As DataGridView, ByVal subject As String)

        Dim counter As Int16 = 1
        resultDGV.Rows.Add(subject)
        For k = 0 To dgv.Rows.Count - 1
            resultDGV.Rows.Add(counter, dgv.Rows(k).Cells(1).Value, dgv.Rows(k).Cells(2).Value, dgv.Rows(k).Cells(3).Value, dgv.Rows(k).Cells(4).Value, dgv.Rows(k).Cells(5).Value, dgv.Rows(k).Cells(6).Value)
            counter += 1
        Next

    End Sub


    Public Function getSchoolSubjects()
        schoolSubjects.Clear()
        Dim available As Boolean = False
        If qread("select abbreviation from subjects;") Then
            If dbreader.RecordsAffected > 0 Then
                available = True
                While dbreader.Read
                    schoolSubjects.Add(dbreader("abbreviation"))
                End While
            End If
        End If

        Return available
    End Function


    Public Sub reorderColumns(ByRef dgv As DataGridView, Optional ByVal delete As Boolean = True)

        If delete Then
            dgv.Rows.Clear()
            dgv.Columns.Clear()
        End If

        dgv.Columns.Add("No", "No")
        dgv.Columns.Add("admno", "Adm No.")
        dgv.Columns.Add("student_name", "Name Of Student")
        dgv.Columns.Add("gender", "Gender")



        dgv.Columns.Add("marks", "marks")
        dgv.Columns.Add("grade", "grade")
        dgv.Columns.Add("point", "Points")

    End Sub

    Private Sub btnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        Me.Close()
    End Sub

    Private Sub btnPrint_Click(sender As Object, e As EventArgs) Handles btnPrint.Click
        If resultDGV.Rows.Count > 0 Then
            generateFromDataTable("Subject Wise Ranking", "From Grid", String.Empty, Nothing, resultDGV)
        Else
            MsgBox("There are no row to print")
        End If
    End Sub
End Class